rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2778 @@
+    type Archetype = Bar;
+}
+```
+
+We can prevent these sorts of silly archetypes by requiring that `Archetype` itself is stable, and that its stability archetype is not transitively defined:
+
+```rust
+pub trait PromiseTransmutableFrom
+{
+    type Archetype
+        : TransmuteInto<Self, NeglectStability>
+        + PromiseTransmutableFrom;
+}
+
+pub trait PromiseTransmutableInto
+{
+    type Archetype
+        : TransmuteFrom<Self, NeglectStability>
+        + PromiseTransmutableInto;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## No Notion of Platform Stability
+The stability declaration traits communicate library layout stability, but not *platform* layout stability. A transmutation is platform-stable if it compiling one one platform implies it will compile on all other platforms. Unfortunately, platform-unstable types are common;  e.g.:
+
+- All pointer-related primitive types (`usize`, `isize`, `*const T`, `*mut T`, `&T`, `&mut T`) possess platform-dependent layouts; their sizes and alignments are well-defined, but vary between platforms. Concretely, whether `usize` is `TransmuteInto<[u8; 4]>` or `TransmuteInto<[u8; 8]>` will depend on  the platform.
+- The very existence of some types depends on platform, too; e.g., the contents of [`core::arch`](https://doc.rust-lang.org/stable/core/arch/), [`std::os`](https://doc.rust-lang.org/stable/std/os/), and [`core::sync::atomic`](https://doc.rust-lang.org/stable/std/sync/atomic/) all depend on platform.
+
+Our proposed stability system is oblivious to the inter-platform variations of these types. Expanding our stability system to be aware of inter-platform variations would introduce considerable additional complexity:
+<ol>
+<li>
+
+**Cognitive Complexity:** For types whose layout varies between platforms, the [stability] declaration traits could, *perhaps*, be adapted to encode platform-related guarantees. We anticipate this would contribute substantial cognitive complexity. Type authors, even those with no interest in cross-platform stability, would nonetheless need to reason about the layout properties of their types on platforms that might not yet exist.
+</li>
+<li>
+
+**Ergonomic Complexity:** Platform instabilities are contagious: a type that *contains* a platform-unstable type is, itself, platform-unstable. Due to the sheer virulence of types with platform-dependent layouts, an explicit '`NeglectPlatformStability`' option would need to be used for *many* simple transmutations. The ergonomic cost of this would also be substantial.
+
+</li>
+<li>
+
+**Implementation Complexity:** The mechanisms proposed by this RFC are, fundamentally, applications of and additions to Rust's type system (i.e., they're traits). Mechanisms that impact platform stability, namely `#[cfg(...)]` annotations, long precede type-resolution and layout computation in the compilation process. For instance, it's possible to define types with impossible layouts:    
+```rust
+#[cfg(any())]
```

Comment:
This is wrong. The Recursive type here is actually never instantiated. `#[cfg(any())]` is equivalent to `#[cfg(false)]`. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e39b13b30504b6c05dfd47dcdbad1603)